homeassistant/components/viam/manager.py

```diff
@@ -0,0 +1,114 @@
+"""Manage Viam client connection."""
+
+import base64
+from typing import Any
+
+from PIL import Image
+from viam.app.app_client import RobotPart
+from viam.app.viam_client import ViamClient
+from viam.robot.client import RobotClient
+from viam.rpc.dial import Credentials, DialOptions
+from viam.services.vision.client import RawImage
+
+from homeassistant.components import camera
+from homeassistant.const import CONF_ADDRESS
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ServiceValidationError
+
+from .const import (
+    CONF_API_ID,
+    CONF_CREDENTIAL_TYPE,
+    CONF_ROBOT_ID,
+    CONF_SECRET,
+    CRED_TYPE_LOCATION_SECRET,
+    DOMAIN,
+)
+
+
+def _fetch_image(filepath: str | None):
+    if filepath is None:
+        return None
+    return Image.open(filepath)
+
+
+class ViamManager:
+    """Manage Viam client and entry data."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        viam: ViamClient,
+        entry_id: str,
+        data: dict[str, Any],
+    ) -> None:
+        """Store initialized client and user input data."""
+        self.hass = hass
+        self.viam = viam
+        self.data = data
+        self.entry_id = entry_id
+
+    def unload(self) -> None:
+        """Clean up any open clients."""
+        self.viam.close()
+
+    def encode_image(self, image: Image.Image | RawImage):
+        """Create base64-encoded Image string."""
+        image_bytes = b""
+        if isinstance(image, Image.Image):
+            image_bytes = image.tobytes()
+        if isinstance(image, RawImage):
+            image_bytes = image.data
+
+        image_string = base64.b64encode(image_bytes).decode()
+        return f"data:image/jpeg;base64,{image_string}"
+
+    async def get_image(self, filepath: str | None, camera_entity: str | None):
+        """Retrieve image type from camera entity or file system."""
+        if filepath is not None:
+            return await self.hass.async_add_executor_job(_fetch_image, filepath)
+        if camera_entity is not None:
+            image = await camera.async_get_image(self.hass, camera_entity)
+            return RawImage(image.content, image.content_type)
+
+        return None
```

Comment:
Should we not require that either filepath or camera_entity is not None?

Reply:
I don't believe so. If it's not possible to get an image when this service is called, I believe it should just return "empty state" values as handled by each service: https://github.com/HipsterBrown/home-assistant-core/blob/3103d540017b9dfca1b364e18179306a83febd44/homeassistant/components/viam/services.py#L237-L241